via/cache_buster.py

```diff
@@ -0,0 +1,57 @@
+"""Pyramid CacheBusters."""
+
+import os
+
+
+class PathCacheBuster:
+    """A cache buster which modifies paths to enable immutable caching.
+
+    Implements `pyramid.interfaces.ICacheBuster` and can be used with
+    `config.add_cache_buster()` to enable rewriting of URLs to a version
+    which contains the last modified time.
+
+    This allows assets to be statically cached until at least one of them has
+    changed. This is achieved by generating a salt value from the last
+    modified time of the files which is applied to the input URL.
+
+    :param path: Path to static for generating the salt value
+    """
+
+    def __init__(self, path):
+        self.salt = self._generate_salt(path)
+
+    def __call__(self, request, subpath, kw):  # pylint: disable=invalid-name
+        """Prepend the salt to the path.
+
+        Implements the ICacheBuster iterface.
+        """
+        return f"{self.salt}/{subpath}", kw
+
+    @property
+    def immutable_path(self):
+        """Return the external URL where immutable content is accessible."""
+        return f"/static/{self.salt}"
+
+    def get_immutable_file_test(self):
+        """Return an immutability test for use with Whitenoise."""
+        immutable_path = self.immutable_path
+
+        def _test(_, url):
+            return url.startswith(immutable_path)
+
+        return _test
+
+    @staticmethod
+    def _generate_salt(path):
+        """Generate salt based on the last modified time.
```

Comment:
Is it safe to use mtimes of files to generate salt? I'm not sure what the mtime really represents. I think Jenkins does a git clone of the repo, at which point the mtimes of all the files will be set to the time when the git clone was ran. Jenkins then builds the Docker image which involves copying all the files into the Docker container, which may or may not change the mtimes again not sure. But anyway, I don't think the mtime of the file on disk is really when the file was last "modified" in a meaningful way?

Should we use a hash of the file's contents instead?

Reply:
It's safe, but it's not as efficient as it could be. 

To be safe we need to ensure that we change every time we make a change in the files. If we change every deploy, then this will meet this requirement.

It's not efficient, because we don't need to invalidate peoples cache unless something has actually changed. So something which reflects the contents is the most robust approach.

Reply:
Re-implemented with MD5